lib/mondrian/olap/result.rb

```diff
@@ -481,8 +478,70 @@ def self.parse_return_fields(result, params)
             end
           end
 
+          if params[:role_name].present?
+            add_role_validation_fields return_fields, sql_options
+          end
+
           [nonempty_columns, return_fields]
         end
+
+        def self.add_sql_attributes(field, dialect:, sql_query:, max_alias_length:, params:)
+          member = field[:member]
+
+          if table_name = (member.try(:getTableName) || member.try(:getMondrianDefExpression).try(:table))
+            field[:quoted_table_name] = dialect.quoteIdentifier(table_name)
+          end
+
+          field[:column_expression] =
+            case field[:type]
+            when :name
+              if member.respond_to? :getNameExp
+                member.getNameExp.getExpression sql_query
+              end
+            when :property
+              if property = member.getProperties.to_a.detect { |p| p.getName == field[:name] }
+                # property.getExp is a protected method therefore
+                # use a workaround to get the value from the field
+                f = property.java_class.declared_field("exp")
+                f.accessible = true
+                if column = f.value(property)
+                  column.getExpression sql_query
+                end
+              end
+            when :name_or_key
+              member.getNameExp&.getExpression(sql_query) || member.getKeyExp&.getExpression(sql_query)
+            else
+              if member.respond_to? :getKeyExp
+                field[:type] = :key
+                member.getKeyExp.getExpression sql_query
+              else
+                field[:type] = :measure
+                column_expression = member.getMondrianDefExpression.getExpression sql_query
+                if params[:group_by]
+                  member.getAggregator.getExpression column_expression
+                else
+                  column_expression
+                end
+              end
+            end
+
+          column_alias = field[:type] == :key ? "#{field[:name]} (Key)" : field[:name]
+          field[:column_alias] = dialect.quoteIdentifier(
+            max_alias_length ? column_alias[0, max_alias_length] : column_alias
+          )
+        end
+
+        def self.add_role_validation_fields(fields, options = {})
+          fields.map { |f| f[:member] }.uniq.each_with_index do |level_or_member, i|
+            next if level_or_member.is_a?(Java::MondrianOlap::Member)
+            current_level = level_or_member
+            loop do
+              fields << {member: current_level, type: :name_or_key, name: "#{i}:#{current_level.getDepth}"}
```

Comment:
I'm wondering that maybe better add some name prefix to distinguish these special return fields. Maybe `_level:`